---
title: "Assignment 11"
author: "Laura Fusi"
date: "2023-05-19"
output: html_document
---

```{r}
# Load libraries
suppressPackageStartupMessages({
  library(GenomicRanges)
  library(epiwraps)
  library(ggplot2)
  library(rGREAT)
  library(AnnotationHub)
  library(ensembldb)
  library(bsseq)
  library(BiocParallel)
  library(edgeR)
  library(DMRcate)
  library(rtracklayer)
  library(sechm)
  library(pheatmap)
  library(viridis)
  library(data.table)
  library(SummarizedExperiment)
  library(limma)
})
# Set the same reproducible random state
set.seed(40)
```

```{r}
bs <- readRDS("bs.rds")
rowRanges(bs)
pData(bs)
```

# Plot a heatmap of the methylation levels of the genes in top 5 DMR regions

## Hint: use getMeth(..., regions=...) and for the heatmap you can adapt the code we used for the promoters

```{r}
# genes
ah <- AnnotationHub()
ensdb <- ah[["AH109336"]]

chr22 <-  GRanges(seqnames=Rle(c("22")), 
                  ranges = IRanges(1, end=195471971))
genesChr22 <- genes(ensdb, columns=c("gene_seq_start", "gene_seq_end", "gene_name"),
                    filter=GRangesFilter(chr22))
seqlevelsStyle(genesChr22) <- "UCSC"
```

```{r}
design <- model.matrix(~Type+Pair, data=pData(bs)) # design matrix

# adapt for methylation data
methdesign <- modelMatrixMeth(design)

seqAnnot <- sequencing.annotate(bs, methdesign, all.cov=TRUE, coef="Typenormal")
dmrcateRes <- dmrcate(seqAnnot, C=2, min.cpgs = 10, pcutoff=0.01)

dmrRanges <- extractRanges(dmrcateRes, genome="hg38")
saveRDS(dmrRanges, "dmr.rds")

DMR.plot(dmrRanges, dmr=1,phen.col=c(rep("red", 3), rep("green",3)), CpGs=bs, genome="hg38")
dmrRangesGenes <- dmrRanges[!is.na(dmrRanges$overlapping.genes)]

# Get the genes within Differentially methylated regions
topIdx <- order(dmrRangesGenes$min_smoothed_fdr)[1:4]
genesDmr <- unlist(tstrsplit(dmrRangesGenes[topIdx]$overlapping.genes, split=", "))
genesDmr <- genesDmr[!is.na(genesDmr)]
dmrGenes <- genesChr22[genesChr22$gene_name %in% genesDmr]
dmrGenes
```

```{r}
metPr <- bsseq::getMeth(bs, 
                        regions=dmrGenes, #promotersChr22[1:100]
                        what="perRegion")
colnames(metPr) <- colnames(bs)
rownames(metPr) <- dmrGenes$gene_name
metPr <- metPr[!is.na(rowSums(metPr)),]
metPr
```

```{r}
# Plot heatmap of the first most diffentially methylated
annotationCol <- as.data.frame(pData(bs)[,c("Type","Pair")])
rownames(annotationCol) <- colnames(metPr)
annotationCol$Pair <- as.factor(annotationCol$Pair)
annotationCol$Type <- relevel(as.factor(annotationCol$Type), "normal")
pheatmap::pheatmap(metPr, 
                   cluster_rows=TRUE,
                   cluster_cols=FALSE,
                   annotation_col=annotationCol,
                   show_rownames = TRUE,
                   color=rocket(10))
```

# Run an enrichment analysis (rGREAT) of the genes within DMRs vs all genes of chromosome 22

```{r}
# only genes within the top 5 DMRs
job <- submitGreatJob(gr=dmrGenes, bg=genesChr22, species="hg38")
res <- getEnrichmentTables(job)
names(res)
bp <- res$`GO Biological Process`
head(bp)
ggplot(head(bp,15), aes(Hyper_Fold_Enrichment, reorder(breakStrings(name, 10, "\n"), Hyper_Adjp_BH), size=Hyper_Foreground_Region_Hits, color=-log10(Hyper_Adjp_BH))) + geom_point() + scale_color_viridis_c(limits = c(0, 5)) + labs(title = "Biological Processes in the Top 5 DMRs of Chromosome 22", x = "Hyper Fold Enrichment", y = "Name of the Biological Process", colour = "Minus Logarithm of the \nAdjusted Hypergeometric p-Values \n(-log10(Hyper_Adjp_BH))", size = "Hyper Foreground Region Hits") + theme(plot.title = element_text(hjust = 0.5))
```

```{r}
# all genes within DMRs
genesDmr_all <- unlist(tstrsplit(dmrRangesGenes$overlapping.genes, split=", "))
genesDmr_all <- genesDmr_all[!is.na(genesDmr_all)]
dmrGenes_all <- genesChr22[genesChr22$gene_name %in% genesDmr_all]
job <- submitGreatJob(gr=dmrGenes_all, bg=genesChr22, species="hg38")
res <- getEnrichmentTables(job)
names(res)
bp <- res$`GO Biological Process`
head(bp)
ggplot(head(bp,15), aes(Hyper_Fold_Enrichment, reorder(breakStrings(name, 10, "\n"), Hyper_Adjp_BH), size=Hyper_Foreground_Region_Hits, color=-log10(Hyper_Adjp_BH))) + geom_point() + scale_color_viridis_c(limits = c(0, 5)) + labs(title = "Biological Processes in all DMRs of Chromosome 22", x = "Hyper Fold Enrichment", y = "Name of the Biological Process", colour = "Minus Logarithm of the \nAdjusted Hypergeometric p-Values \n(-log10(Hyper_Adjp_BH))", size = "Hyper Foreground Region Hits") + theme(plot.title = element_text(hjust = 0.5))
```

## Describe what your enrichment analysis is testing exactly & describe the results

I have performed two different enrichment analyses testing two slightly different things. In both cases I get a table showing for each biological function how much it is represented in the selected regions compared to all genes of chromosome 22, and how significant this difference is, and a plot, to visualize the results. The most important columns of the table basically tell for each biological process how many regions you can identify in those DMRs with that function, how many you would normally expect just by chance in chromosome 22, the ratio between them, and the p-value (the raw one and the adjusted one). This is more or less also what the plot represents. For each biological function (represented on the y-axis), we can see the hyper fold enrichment on the x-axis, which is the the proportion of genes related to that function compared to the rest of the genes (in the selected set of DMRs). In addition, the size of the dot represents whether this is a lot or not (compared to chromosome 22), and the color represents whether these findings are significant (yellow stands for a lower p-value (more significant), whereas darker colours represent higher p-values). If we assume a significance threshold of 5%, then -log10(0.05) = 1.3, therefore all the dots that tend to be from dark green to yellow are significant, whereas the dots that are blue/ dark blue/ dark violet are not significant. 

Firstly, I have tested only the top 5 DMRs vs all genes of chromosome 22. There all the adjusted p-values seems to be significant, meaning that the given biological functions are significantly more represented in this set of DMRs compared with the genes of chromosome 22. One function that particularly catches one's eye is "cell-cell adhesion via plasma-membrane adhesion molecules". The hyper fold enrichment is not that large, if compared with the other functions, but the dot is really large, meaning that the difference is still very important. Indeed, it is well known that cancer cells change their "adhesion behaviour".

Secondly, I have tested all DMRs vs all genes of chromosome 22. There we the hyperfold enrichment for the different biological functions is lower compared to the previous analysis, but the differences we catch are much more evident and significant. This is probably due to the fact that we are looking at a broader set of DMRs, while always comparing with the genes in chromosome 22. Some of the listed functions are quite vague, but others that are more clear are very interesting, e.g."cellular component organization or biogenesis". Some of these functions are known to be changed in cancer cells.

Overall, this analysis shows that some biological processes that are important/ changed in cancer development seem to be associated with DNA methylation regulation.

