---
title: "Assignment 10"
author: "Laura Fusi"
date: "2023-05-12"
output: html_document
---

```{r}
suppressPackageStartupMessages({
  library(GenomicRanges)
  library(epiwraps)
  library(ggplot2)
  library(rGREAT)
})
```

# Download and decompress the following archive
* https://ethz-ins.org/content/w10.assignment.zip
* This contains the bigwig files and peaks (bed) files for three TFs of the CREB family (all restricted to chr1; aligned against the hg38 genome)

```{r, eval=FALSE}
options(timeout = 6000)
download.file("https://ethz-ins.org/content/w10.assignment.zip", "w10.assignment.zip")
unzip("w10.assignment.zip")
list.files()
```

# Use clustering and visualization to illustrate the relationship between the binding of the different proteins

```{r}
# prepare regions and tracks
tracks <- list.files(pattern="bw$")
region_files <- list.files(pattern="bed$")
peaks <- lapply(region_files, FUN=rtracklayer::import.bed)
peaks <- lapply(peaks, FUN=function(x) x[x$score>800])
regions <- reduce(unlist(GRangesList(peaks)))
# plot
ml <- signal2Matrix(tracks, regions, extend=2000)
set.seed(123)  
nf <- getNormFactors(tracks, useSeqLevels="chr1")
ml2 <- rescaleSignalMatrices(ml, scaleFactors=nf)
plotEnrichedHeatmaps(ml)
# Choosing a meaningful number of clusters
multi_cl <- clusterSignalMatrices(ml, k=2:10)
ggplot(multi_cl$varExplained, aes(k, varExplained)) + geom_line()
# Clustering
set.seed(123)  # to ensure that it gives the same results everytime
cl <- clusterSignalMatrices(ml, k=6) # choose k according to previous plot
table(cl)
head(cl)
length(cl)
length(regions)
regions$cluster <- cl
# Plot clusters with colors
mycolors <- c("1"="red", "2"="blue", "3"="darkgreen", "4"="black", "5"="orange", "6"="darkviolet")
plotEnrichedHeatmaps(ml, row_split=cl, mean_color=mycolors)
# Plotting averages
d <- meltSignals(ml, splitBy=cl)
ggplot(d, aes(position, mean, colour=sample)) + geom_line(size=1.2) + facet_wrap(~split)
```

# Use enrichment analysis (either GO or motif) on at least one of the clusters

```{r}
#' breakStrings
#'
#' breaks a string of words (or vector thereof) into two lines
#'
#' @param x a character vector
#' @param minSizeForBreak the minimum number of characters to break on two lines (default 20)
#' @param lb the line break character (default "\n")
#'
#' @return a character vector of length=length(x)
#'
#' @export
breakStrings <- function(x, minSizeForBreak=20, lb="\n"){
   sapply(x,minSizeForBreak=minSizeForBreak,lb=lb,FUN=function(x,minSizeForBreak,lb){
      if(nchar(x)<=minSizeForBreak) return(x)
      g <- gregexpr(" ", x)[[1]]
      if(length(g)==0) return(x)
      if(length(g)==1 & all(g==-1)) return(x)
      mid <- nchar(x)/2
      mid <- g[order(abs(g-mid))[1]]
      substr(x, mid, mid) <- lb
      return(x)
   })
}
```


```{r, fig.width=10, fig.height=6}
# Split the regions by cluster
split_regions <- split(regions, cl)
# Send the data to the GREAT server to run an enrichment analysis
job <- submitGreatJob(gr=split_regions[["3"]], bg=regions, species="hg38")
res <- getEnrichmentTables(job)
names(res)
# Biological process
bp <- res$`GO Biological Process`
head(bp)
ggplot(head(bp,15), aes(Hyper_Fold_Enrichment, reorder(breakStrings(name, 10, "\n"), Hyper_Adjp_BH), size=Hyper_Foreground_Region_Hits, color=-log10(Hyper_Adjp_BH))) + geom_point() + scale_color_viridis_c(limits = c(0, 4.65)) + labs(title = "Biological Processes in the 3rd Cluster", x = "Hyper Fold Enrichment", y = "Name of the Biological Process", colour = "Minus Logarithm of the \nAdjusted Hypergeometric p-Values \n(-log10(Hyper_Adjp_BH))", size = "Hyper Foreground Region Hits") + theme(plot.title = element_text(hjust = 0.5))
```

# Write a paragraph describing your results

If we look at the enrichment table we get, it has several columns. The most important columns basically tell for each biological process how many regions you can identify in that cluster with that function, how many you would normally expect just by chance, the ratio between them, and the p-value (the raw one and the adjusted one). This is more or less also what the plot represents. For each biological function (represented on the y-axis), we can see the hyper fold enrichment on the x-axis, which is the the proportion of genes related to that function compared to the rest of the genes. In addition, the size of the dot represents whether this is a lot or not, and the color represents whether these findings are significant (yellow stands for a lower p-value (more significant), whereas darker colours represent higher p-values). As we can see the most significant result we have is the one for "response to stimulus", but the significance of the other findings is not that bad. If we assume a significance threshold of 5%, then -log10(0.05) = 1.3, therefore all the dots that tend to be from dark green to yellow are significant, whereas the dots that are blue/ dark blue/ dark violet are not significant. It is not that much surprising that "response to stimulus" is the most significant result, because from the heat map we can see that the intensity for Creb3 is very high in this cluster.
