---
title: "Assignment 6"
author: "Laura Fusi"
date: "2023-03-31"
output: html_document
---

```{r}
suppressPackageStartupMessages({
  library(AnnotationHub)
  library(GenomicRanges)
  library(ggplot2)
  # library(memes) # for the meme-based methods -- COMMENT OUT when using alternatives
  library(motifmatchr)
  library(Biostrings) # for handling sequences
  library(MotifDb) # database of motifs
  library(TFBSTools) # for handling some motif formats
  library(universalmotif) # for converting motifs to various formats
  library(PWMEnrich) # for R-based motif enrichment analysis
})
```

# Choose a transcription factor, e.g. CREB1, REST, GATA5, EGR1, GCR
(or any of your choice that has a motif and available ChIPseq data)

--> REST

# Download the (e.g. Mouse) peaks for that factor (whatever cell type)

--> Download mouse myocyte peaks for REST

```{r}
download.file("https://www.encodeproject.org/files/ENCFF368VWJ/@@download/ENCFF368VWJ.bed.gz", "myo_REST.bed.gz")
REST_peaks <- rtracklayer::import("myo_REST.bed.gz", format="narrowPeak")
seqlevelsStyle(REST_peaks) <- "Ensembl"
peaks_chr1 <- REST_peaks[seqnames(REST_peaks)=="1"]
```

# Identify the instances of the factorâ€™s motif

```{r}
# Obtain peak sequences
peak_centers <- resize(REST_peaks, fix="center", width=100)
ah <- AnnotationHub(localHub=FALSE)
genome <- ah[["AH68356"]]
peak_seqs <- Biostrings::getSeq(genome, peak_centers)
names(peak_seqs) <- as.character(granges(peak_centers))
peak_seqs
```
```{r}
# Getting the desired motif and visualize it
motifs <- query(MotifDb, "REST")
names(motifs)
motif <- motifs[["Mmusculus-HOCOMOCOv10-REST_MOUSE.H10MO.A"]]
view_motifs(motif)
```
```{r}
# Function doing something similar to matchMotifs
#' findMotifInstances
#'
#' A wrapper around TFBSTools for scanning motif occurence, handling the 
#' coordinate conversion as `memes` does.
#' 
#' @param seqs A set of sequences, e.g. `DNAStringSet`, optionally with 
#'   coordinate as names to enable conversion.
#' @param motif A motif, in any format recognized by `universalmotif`
#' @param keepMatchedSeq Logical; whether to keep the matched sequence.
#' @param ... Passed to `TFBSTools::searchSeq`; can for instance be used to set
#'   the number of threads to use, e.g. with `mc.cores=2`
#'
#' @return A `GRanges` object
findMotifInstances <- function(seqs, motif, keepMatchedSeq=FALSE, ...){
  motif <- universalmotif::convert_motifs(motif, "TFBSTools-PWMatrix")
  x <- suppressWarnings(TFBSTools::searchSeq(motif, subject=seqs))
  peaks <- strsplit(gsub("-",":",names(seqs)), ":")
  if(all(lengths(peaks)==3)){ # convert relative coordinates to absolute
    chrs <- sapply(peaks,FUN=function(x) x[1])
    offsets <- sapply(peaks,FUN=function(x) as.integer(x[2]))-1L
    i <- rep(seq_along(x),lengths(x))
    return(GRanges(chrs[i], 
                   IRanges(offsets[i]+as.integer(unlist(lapply(x, start))),
                           offsets[i]+as.integer(unlist(lapply(x, end)))),
                   strand=unlist(lapply(x, strand)),
                   score=as.numeric(unlist(lapply(x, FUN=function(x) score(x))))
                   ))
  }
  x <- as(x, "GRanges")
  score(x) <- x$absScore
  keepFields <- c("score","relScore")
  if(keepMatchedSeq) keepFields <- c(keepFields, "siteSeqs")
  mcols(x) <- mcols(x)[,intersect(colnames(mcols(x)), keepFields)]
  x
}
```

```{r}
# Scanning the peak sequences
Biostrings::writeXStringSet(rtracklayer::import(genome), "genome.fa")
motif2 <- convert_motifs(motif, class="TFBSTools-PFMatrix")
# moi2 <- matchMotifs(motif2, subject=peak_centers, genome=Rsamtools::FaFile("genome.fa"), out="positions")
moi2 <- findMotifInstances(motif = motif2, seqs = peak_seqs)
moi2
```

```{r}
# Scanning the entire genome sequence
Biostrings::writeXStringSet(rtracklayer::import(genome), "genome.fa")
motifg <- convert_motifs(motif, class="TFBSTools-PFMatrix")
moig <- findMotifInstances(motif = motifg, seqs = Biostrings::getSeq(genome))
moig
```

# Answer the following questions
## Of all the peaks, what proportion contains a motif for the factor?
(Expected form of an answer: of the XX peaks, XX (XX%) contain a motif)

```{r}
PeaksOverlapsMotifs <- overlapsAny(peak_centers,moi2)
table(PeaksOverlapsMotifs)
table(PeaksOverlapsMotifs)[2]/(table(PeaksOverlapsMotifs)[1]+table(PeaksOverlapsMotifs)[2])
```
```{r}
length(peak_centers)
```
Of the 3555 peaks, 1722 (48%) contain a motif. There the motif instances where calculated only in the peak regions.
 
```{r}
PeaksOverlapsMotifsG <- overlapsAny(peak_centers,moig)
table(PeaksOverlapsMotifsG)
table(PeaksOverlapsMotifsG)[2]/(table(PeaksOverlapsMotifsG)[1]+table(PeaksOverlapsMotifsG)[2])
```

```{r}
length(peak_centers)
```
 Of the 3555 peaks, 1802 (51%) contain a motif. There the motif instances were calculated on the whole genome.
 
 The results do not differ that much there.
 
## Of all instances of that motif in the genome (or in one chromosome), what proportion is bound by the factor (i.e. has a peak)?
(Expected form of an answer: of the XX motif instances, XX (XX%) overlap a peak)

```{r}
MotifsOverlapsPeaks <- overlapsAny(moi2,peak_centers)
table(MotifsOverlapsPeaks)
table(MotifsOverlapsPeaks)[2]/(table(MotifsOverlapsPeaks)[1]+table(MotifsOverlapsPeaks)[2])
```
```{r}
length(moi2)
```
All 1850 motifs (100%) in the peak regions are bound by the factor. There the motif instances where calculated only in the peak regions, so it makes sense that using these for the overlap calculation we obtain 100%.

```{r}
MotifsOverlapsPeaksG <- overlapsAny(moig,peak_centers)
table(MotifsOverlapsPeaksG)
table(MotifsOverlapsPeaksG)[2]/(table(MotifsOverlapsPeaksG)[1]+table(MotifsOverlapsPeaksG)[2])
```
```{r}
length(moig)
```
Of the 209697 motif instances in the genome, 1992 (0.95%) overlap a peak. There the motif instances were calculated on the whole genome, so it makes sense that there are not that many overlaps in %, and it is more interesting to know this number than the previous one.

```{r}
# Just to check whether I get the same result with this different formulation
length(moig[overlapsAny(moig,peak_centers)])/length(moig)
```